File: packages/polycentric-core/src/api-methods.ts
import * as Base64 from '@borderless/base64';
import fetch, { Headers } from 'cross-fetch';
import Long from 'long';
import * as Models from './models';
import * as Protocol from './protocol';
import * as Version from './version';

async function checkResponse(name: string, response: Response): Promise<void> {
  if (!response.ok) {
    console.warn(name, response.status, await response.text());
    throw new Error(name + ' !ok');
  }
}

function encodeModerationLevels(
  moderationLevels: Record<string, number>,
): string {
  return JSON.stringify(
    Object.entries(moderationLevels).map(([key, value]) => ({
      name: key,
      max_level: value,
      strict_mode: false,
    })),
  );
}

const userAgent = 'polycentric-core-' + Version.SHA.substring(0, 8);

export async function postEvents(
  server: string,
  events: Models.SignedEvent.SignedEvent[],
): Promise<void> {
  const response = await fetch(server + '/events', {
    method: 'POST',
    headers: new Headers({
      'x-polycentric-user-agent': userAgent,
    }),
    body: Protocol.Events.encode({
      events: events,
    }).finish(),
  });

  await checkResponse('postEvents', response);
}

export async function postCensor(
  server: string,
  censorshipType: string,
  urlInfo: string,
  authorization: string,
): Promise<void> {
  const response = await fetch(
    `${server}/censor?censorship_type=${censorshipType}`,
    {
      method: 'POST',
      headers: new Headers({
        authorization: authorization,
        'x-polycentric-user-agent': userAgent,
      }),
      body: urlInfo,
    },
  );

  await checkResponse('postCensor', response);
}

export async function getRanges(
  server: string,
  system: Models.PublicKey.PublicKey,
): Promise<Models.Ranges.RangesForSystem> {
  const systemQuery = Base64.encodeUrl(
    Protocol.PublicKey.encode(system).finish(),
  );

  const path = `/ranges?system=${systemQuery}`;

  const response = await fetch(server + path, {
    method: 'GET',
    headers: new Headers({
      'x-polycentric-user-agent': userAgent,
    }),
  });

  await checkResponse('getRanges', response);

  const rawBody = new Uint8Array(await response.arrayBuffer());

  return Models.Ranges.rangesForSystemFromBuffer(rawBody);
}

export type GetEventsType = (
  server: string,
  system: Models.PublicKey.PublicKey,
  ranges: Models.Ranges.RangesForSystem,
  moderationLevels?: Record<string, number>,
) => Promise<Models.Events.Type>;

export const getEvents: GetEventsType = async (
  server: string,
  system: Models.PublicKey.PublicKey,
  ranges: Models.Ranges.RangesForSystem,
  moderationLevels?: Record<string, number>,
): Promise<Models.Events.Type> => {
  const systemQuery = Base64.encodeUrl(
    Protocol.PublicKey.encode(system).finish(),
  );

  const rangesQuery = Base64.encodeUrl(
    Protocol.RangesForSystem.encode(ranges).finish(),
  );

  let path = `/events?system=${systemQuery}&ranges=${rangesQuery}`;

  if (moderationLevels !== undefined) {
    const moderationLevelsQuery = encodeModerationLevels(moderationLevels);
    path += `&moderation_filters=${moderationLevelsQuery}`;
  }

  const response = await fetch(server + path, {
    method: 'GET',
    headers: new Headers({
      'x-polycentric-user-agent': userAgent,
    }),
  });

  await checkResponse('getEvents', response);

  const rawBody = new Uint8Array(await response.arrayBuffer());

  return Models.Events.fromBuffer(rawBody);
};

export async function getResolveClaim(
  server: string,
  trustRoot: Models.PublicKey.PublicKey,
  claimType: Models.ClaimType.ClaimType,
  matchAnyField: string,
): Promise<Models.QueryClaimToSystemResponse.ResponseType> {
  const query = Base64.encodeUrl(
    Protocol.QueryClaimToSystemRequest.encode({
      claimType: claimType,
      trustRoot: trustRoot,
      matchAnyField: matchAnyField,
    }).finish(),
  );

  const path = `/resolve_claim?query=${query}`;

  const response = await fetch(server + path, {
    method: 'GET',
    headers: new Headers({
      'x-polycentric-user-agent': userAgent,
    }),
  });

  await checkResponse('getResolveClaim', response);

  const rawBody = new Uint8Array(await response.arrayBuffer());

  return Models.QueryClaimToSystemResponse.responseTypeFromBuffer(rawBody);
}

export type GetQueryLatestType = (
  server: string,
  system: Models.PublicKey.PublicKey,
  eventTypes: Models.ContentType.ContentType[],
) => Promise<Models.Events.Type>;

export const getQueryLatest: GetQueryLatestType = async (
  server: string,
  system: Models.PublicKey.PublicKey,
  eventTypes: Models.ContentType.ContentType[],
): Promise<Models.Events.Type> => {
  const systemQuery = Base64.encodeUrl(
    Protocol.PublicKey.encode(system).finish(),
  );

  const eventTypesQuery = Base64.encodeUrl(
    Protocol.RepeatedUInt64.encode({
      numbers: eventTypes,
    }).finish(),
  );

  const path =
    `/query_latest?system=${systemQuery}` + `&event_types=${eventTypesQuery}`;

  const response = await fetch(server + path, {
    method: 'GET',
    headers: new Headers({
      'x-polycentric-user-agent': userAgent,
    }),
  });

  await checkResponse('getQueryLatest', response);

  const rawBody = new Uint8Array(await response.arrayBuffer());

  return Models.Events.fromBuffer(rawBody);
};

export async function getQueryIndex(
  server: string,
  system: Models.PublicKey.PublicKey,
  contentType: Models.ContentType.ContentType,
  after?: Long,
  limit?: Long,
): Promise<Models.QueryIndexResponse.Type> {
  const systemQuery = Base64.encodeUrl(
    Protocol.PublicKey.encode(system).finish(),
  );

  const path =
    `/query_index?system=${systemQuery}` +
    `&content_type=${contentType.toString()}` +
    (after ? `&after=${after.toString()}` : '') +
    (limit ? `&limit=${limit.toString()}` : '');

  const response = await fetch(server + path, {
    method: 'GET',
    headers: new Headers({
      'x-polycentric-user-agent': userAgent,
    }),
  });

  await checkResponse('getQueryIndex', response);

  const rawBody = new Uint8Array(await response.arrayBuffer());

  return Models.QueryIndexResponse.fromBuffer(rawBody);
}

export async function getQueryReferences(
  server: string,
  reference: Protocol.Reference,
  cursor?: Uint8Array,
  requestEvents?: Protocol.QueryReferencesRequestEvents,
  countLwwElementReferences?: Protocol.QueryReferencesRequestCountLWWElementReferences[],
  countReferences?: Protocol.QueryReferencesRequestCountReferences[],
  extraByteReferences?: Uint8Array[],
  moderationLevels?: Record<string, number>,
): Promise<Protocol.QueryReferencesResponse> {
  const query: Protocol.QueryReferencesRequest = {
    reference: reference,
    cursor: cursor,
    requestEvents: requestEvents,
    countLwwElementReferences: countLwwElementReferences ?? [],
    countReferences: countReferences ?? [],
    extraByteReferences: extraByteReferences ?? [],
  };

  const encodedQuery = Base64.encodeUrl(
    Protocol.QueryReferencesRequest.encode(query).finish(),
  );

  let path = `/query_references?query=${encodedQuery}`;

  if (moderationLevels !== undefined) {
    path += `&moderation_filters=${encodeModerationLevels(moderationLevels)}`;
  }

  const response = await fetch(server + path, {
    method: 'GET',
    headers: new Headers({
      'x-polycentric-user-agent': userAgent,
    }),
  });

  await checkResponse('getQueryReferences', response);

  const rawBody = new Uint8Array(await response.arrayBuffer());

  return Protocol.QueryReferencesResponse.decode(rawBody);
}

export enum SearchType {
  Messages = 'messages',
  Profiles = 'profiles',
}

export async function getSearch(
  server: string,
  searchQuery: string,
  limit?: number,
  cursor?: Uint8Array,
  searchType?: SearchType,
  moderationLevels?: Record<string, number>,
): Promise<Models.ResultEventsAndRelatedEventsAndCursor.Type> {
  let path = `/search?search=${encodeURIComponent(searchQuery)}`;

  if (cursor !== undefined) {
    path += `&cursor=${Base64.encodeUrl(cursor)}`;
  }

  if (limit !== undefined) {
    path += `&limit=${limit.toString()}`;
  }

  if (searchType !== undefined) {
    path += `&search_type=${searchType}`;
  }

  if (moderationLevels !== undefined) {
    path += `&moderation_filters=${encodeModerationLevels(moderationLevels)}`;
  }

  const response = await fetch(server + path, {
    method: 'GET',
    headers: new Headers({
      'x-polycentric-user-agent': userAgent,
    }),
  });

  await checkResponse('getSearch', response);

  const rawBody = new Uint8Array(await response.arrayBuffer());

  return Models.ResultEventsAndRelatedEventsAndCursor.fromBuffer(rawBody);
}

export type TopStringReferenceTimeRange = '12h' | '1d' | '7d' | '30d';

export async function getTopStringReferences(
  server: string,
  options: {
    query?: string;
    timeRange?: TopStringReferenceTimeRange;
    limit?: number;
  },
): Promise<Models.ResultTopStringReferences.Type> {
  let path = '/top_string_references?';

  const params = new URLSearchParams();

  if (options.query !== undefined) {
    params.append('query', options.query);
  }

  if (options.limit !== undefined) {
    params.append('limit', options.limit.toString());
  }

  if (options.timeRange !== undefined) {
    params.append('time_range', options.timeRange);
  }

  path += params.toString();

  const response = await fetch(server + path, {
    method: 'GET',
    headers: new Headers({
      'x-polycentric-user-agent': userAgent,
    }),
  });

  await checkResponse('getTopStringReferences', response);

  const rawBody = new Uint8Array(await response.arrayBuffer());

  return Models.ResultTopStringReferences.fromBuffer(rawBody);
}

export async function getHead(
  server: string,
  system: Models.PublicKey.PublicKey,
): Promise<Models.Events.Type> {
  const systemQuery = Base64.encodeUrl(
    Protocol.PublicKey.encode(system).finish(),
  );

  const path = `/head?system=${systemQuery}`;

  const response = await fetch(server + path, {
    method: 'GET',
    headers: new Headers({
      'x-polycentric-user-agent': userAgent,
    }),
  });

  await checkResponse('getHead', response);

  const rawBody = new Uint8Array(await response.arrayBuffer());

  return Models.Events.fromBuffer(rawBody);
}

export async function getExplore(
  server: string,
  limit?: number,
  cursor?: Uint8Array,
  moderationLevels?: Record<string, number>,
): Promise<Models.ResultEventsAndRelatedEventsAndCursor.Type> {
  let path = '/explore?';

  const params = new URLSearchParams();

  if (cursor !== undefined) {
    params.append('cursor', Base64.encodeUrl(cursor));
  }

  if (limit !== undefined) {
    params.append('limit', limit.toString());
  }

  if (moderationLevels !== undefined) {
    params.append(
      'moderation_filters',
      encodeModerationLevels(moderationLevels),
    );
  }

  path += params.toString();

  const response = await fetch(server + path, {
    method: 'GET',
    headers: new Headers({
      'x-polycentric-user-agent': userAgent,
    }),
  });

  await checkResponse('getExplore', response);

  const rawBody = new Uint8Array(await response.arrayBuffer());

  return Models.ResultEventsAndRelatedEventsAndCursor.fromBuffer(rawBody);
}

export async function getFindClaimAndVouch(
  server: string,
  vouching_system: Models.PublicKey.PublicKey,
  claiming_system: Models.PublicKey.PublicKey,
  fields: Protocol.ClaimFieldEntry[],
  claimType: Models.ClaimType.ClaimType,
): Promise<Models.FindClaimAndVouchResponse.Type | undefined> {
  const query: Protocol.FindClaimAndVouchRequest = {
    vouchingSystem: vouching_system,
    claimingSystem: claiming_system,
    fields: fields,
    claimType: claimType,
  };

  const encodedQuery = Base64.encodeUrl(
    Protocol.FindClaimAndVouchRequest.encode(query).finish(),
  );

  const path = `/find_claim_and_vouch?query=${encodedQuery}`;

  const response = await fetch(server + path, {
    method: 'GET',
    headers: new Headers({
      'x-polycentric-user-agent': userAgent,
    }),
  });

  if (response.status === 404) {
    return undefined;
  }

  await checkResponse('getFindClaimAndVouch', response);

  const rawBody = new Uint8Array(await response.arrayBuffer());

  return Models.FindClaimAndVouchResponse.fromBuffer(rawBody);
}

export async function getChallenge(
  server: string,
): Promise<Protocol.HarborChallengeResponse> {
  const response = await fetch(server + '/challenge', {
    method: 'GET',
    headers: new Headers({
      'x-polycentric-user-agent': userAgent,
    }),
  });

  await checkResponse('getChallenge', response);

  const rawBody = new Uint8Array(await response.arrayBuffer());

  return Protocol.HarborChallengeResponse.decode(rawBody);
}

export async function postPurge(
  server: string,
  solvedChallenge: Protocol.HarborValidateRequest,
): Promise<void> {
  const response = await fetch(server + '/purge', {
    method: 'POST',
    body: Protocol.HarborValidateRequest.encode(solvedChallenge).finish(),
    headers: new Headers({
      'x-polycentric-user-agent': userAgent,
    }),
  });

  await checkResponse('postPurge', response);
}

export async function postClaimHandle(
  server: string,
  claimRequest: Protocol.ClaimHandleRequest,
): Promise<void> {
  const response = await fetch(server + '/claim_handle', {
    method: 'POST',
    headers: new Headers({
      'content-type': 'application/octet-stream',
      'x-polycentric-user-agent': userAgent,
    }),
    body: Protocol.ClaimHandleRequest.encode(claimRequest).finish(),
  });

  await checkResponse('postClaimHandle', response);
}

export async function getResolveHandle(
  server: string,
  handle: string,
): Promise<Models.PublicKey.PublicKey> {
  const response = await fetch(server + `/resolve_handle?handle=${handle}`, {
    method: 'GET',
    headers: new Headers({
      'content-type': 'application/octet-stream',
      'x-polycentric-user-agent': userAgent,
    }),
  });

  await checkResponse('getResolveHandle', response);

  const rawBody = new Uint8Array(await response.arrayBuffer());

  return Models.PublicKey.fromProto(Protocol.PublicKey.decode(rawBody));
}

export const VERIFIER_SERVER =
  // Check if we're in a browser environment
  typeof window !== 'undefined'
    ? window.location.hostname === 'localhost' ||
      window.location.hostname === '127.0.0.1'
      ? 'https://localhost:3002' // Local development
      : 'https://verifiers.polycentric.io' // Production
    : process.env.NEXT_PUBLIC_VERIFIER_SERVER ??
      'https://verifiers.polycentric.io';

export async function requestVerification(
  pointer: Protocol.Pointer,
  claimType: Models.ClaimType.ClaimType,
  challengeResponse?: string,
): Promise<void> {
  const verifierType = challengeResponse ? 'oauth' : 'text';

  let url = `${VERIFIER_SERVER}/platforms/${claimType.toString()}/${verifierType}/vouch`;

  if (challengeResponse) {
    url += `?challengeResponse=${encodeURIComponent(challengeResponse)}`;
  }

  try {
    const encodedPointer = Protocol.Pointer.encode(pointer).finish();

    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'content-type': 'application/octet-stream',
        'x-polycentric-user-agent': userAgent,
        Origin: window.location.origin,
        'Access-Control-Request-Method': 'POST',
        'Access-Control-Request-Headers':
          'content-type,x-polycentric-user-agent',
      },
      body: encodedPointer,
      credentials: 'include',
      mode: 'cors',
    });

    if (!response.ok) {
      const text = await response.text();
      console.error('Error response body:', text);
      throw new Error(`Verification failed: ${text}`);
    }

    await checkResponse('requestVerification', response);
  } catch (error) {
    if (error instanceof TypeError && error.message === 'Failed to fetch') {
      console.error('CORS or network error:', {
        error,
        url,
        origin: window.location.origin,
      });
    }
    throw error;
  }
}

export interface OAuthUsernameResponse {
  username: string;
  token: string;
}

interface OAuthURLResponse {
  url: string;
}

export async function getOAuthURL(
  server: string,
  claimType: Models.ClaimType.ClaimType,
  redirectUri?: string,
): Promise<string> {
  let url = `${server}/platforms/${claimType.toString()}/oauth/url`;

  if (redirectUri) {
    url += `?redirectUri=${encodeURIComponent(redirectUri)}`;
  }

  const response = await fetch(url, {
    method: 'GET',
    headers: new Headers({
      'x-polycentric-user-agent': userAgent,
    }),
    credentials: 'include',
    mode: 'cors',
  });

  await checkResponse('getOAuthURL', response);
  const data = (await response.json()) as string | OAuthURLResponse;

  return typeof data === 'string' ? data : data.url;
}

export async function getOAuthUsername(
  server: string,
  token: string,
  claimType: Models.ClaimType.ClaimType,
): Promise<OAuthUsernameResponse> {
  const url = `${server}/platforms/${claimType.toString()}/oauth/token${
    token.startsWith('?') ? token : `?${token}`
  }`;

  const response = await fetch(url, {
    method: 'GET',
    headers: new Headers({
      'x-polycentric-user-agent': userAgent,
    }),
  });

  await checkResponse('getOAuthUsername', response);
  return response.json() as Promise<OAuthUsernameResponse>;
}

export async function getClaimFieldsByUrl(
  server: string,
  claimType: Models.ClaimType.ClaimType,
  subject: string,
): Promise<Protocol.ClaimFieldEntry[]> {
  const url = `${server}/platforms/${claimType.toString()}/text/getClaimFieldsByUrl`;

  const response = await fetch(url, {
    method: 'POST',
    headers: new Headers({
      'Content-Type': 'application/json',
      'x-polycentric-user-agent': userAgent,
    }),
    body: JSON.stringify({ url: subject }),
  });

  await checkResponse('getClaimFieldsByUrl', response);

  interface ClaimFieldItem {
    key: number;
    value: string;
  }

  const decoded = (await response.json()) as ClaimFieldItem[];
  return decoded.map((item: ClaimFieldItem) => ({
    key: Long.fromNumber(item.key),
    value: item.value,
  }));
}

File: client-lib/src/api_methods.rs
#![allow(dead_code)]

use anyhow::{anyhow, Result};
use protobuf::Message;
use reqwest::{Client, RequestBuilder};

use polycentric_protocol::protocol::{
    ClaimFieldEntry, ClaimHandleRequest, Events, FindClaimAndVouchRequest,
    FindClaimAndVouchResponse, HarborChallengeResponse, HarborValidateRequest,
    PublicKey, QueryClaimToSystemRequest, QueryClaimToSystemResponse,
    QueryIndexResponse, QueryReferencesRequest,
    QueryReferencesRequestCountLWWElementReferences,
    QueryReferencesRequestCountReferences, QueryReferencesRequestEvents,
    QueryReferencesResponse, RangesForSystem, Reference, RepeatedUInt64,
};

use polycentric_protocol::model::signed_event::SignedEvent;

pub struct ApiMethods {
    user_agent: String,
    client: Client,
}

impl ApiMethods {
    pub fn new() -> Self {
        ApiMethods {
            user_agent: format!(
                "polycentric-client-{}",
                env!("CARGO_PKG_VERSION")
            ),
            client: Client::new(),
        }
    }

    fn get_request_builder(&self, url: &str) -> RequestBuilder {
        self.client
            .get(url)
            .header("x-polycentric-user-agent", &self.user_agent)
    }

    fn post_request_builder(&self, url: &str) -> RequestBuilder {
        self.client
            .post(url)
            .header("x-polycentric-user-agent", &self.user_agent)
    }

    async fn check_response(
        name: &str,
        response: &reqwest::Response,
    ) -> Result<()> {
        if !response.status().is_success() {
            let status = response.status();
            let url = response.url().clone();
            return Err(anyhow!(
                "{} to {} failed with status code {}.",
                name,
                url,
                status
            ));
        }
        Ok(())
    }

    pub async fn post_events(
        &self,
        server: &str,
        events: &[SignedEvent],
    ) -> Result<()> {
        let body = Events {
            events: events
                .iter()
                .map(polycentric_protocol::model::signed_event::to_proto)
                .collect(),
            ..Default::default()
        }
        .write_to_bytes()?;

        let response = self
            .post_request_builder(&format!("{}/events", server))
            .body(body)
            .send()
            .await?;

        Self::check_response("postEvents", &response).await
    }

    pub async fn post_censor(
        &self,
        server: &str,
        censorship_type: &str,
        url_info: &str,
        authorization: &str,
    ) -> Result<()> {
        let response = self
            .post_request_builder(&format!(
                "{}/censor?censorship_type={}",
                server, censorship_type
            ))
            .header("authorization", authorization)
            .body(url_info.to_string())
            .send()
            .await?;

        Self::check_response("postCensor", &response).await
    }

    pub async fn get_ranges(
        &self,
        server: &str,
        system: &PublicKey,
    ) -> Result<RangesForSystem> {
        let system_query =
            base64::encode_config(system.write_to_bytes()?, base64::URL_SAFE);
        let request = self.get_request_builder(&format!(
            "{}/ranges?system={}",
            server, system_query
        ));

        let response = request.send().await?;
        Self::check_response("getRanges", &response).await?;

        let ranges =
            RangesForSystem::parse_from_bytes(&response.bytes().await?)?;
        Ok(ranges)
    }

    pub async fn get_events(
        &self,
        server: &str,
        system: &PublicKey,
        ranges: &RangesForSystem,
    ) -> Result<Events> {
        let system_query =
            base64::encode_config(system.write_to_bytes()?, base64::URL_SAFE);
        let ranges_query =
            base64::encode_config(ranges.write_to_bytes()?, base64::URL_SAFE);
        let request = self.get_request_builder(&format!(
            "{}/events?system={}&ranges={}",
            server, system_query, ranges_query
        ));

        let response = request.send().await?;
        Self::check_response("getEvents", &response).await?;

        let events = Events::parse_from_bytes(&response.bytes().await?)?;
        Ok(events)
    }

    pub async fn get_resolve_claim(
        &self,
        server: &str,
        trust_root: &PublicKey,
        claim_type: u64,
        match_any_field: &str,
    ) -> Result<QueryClaimToSystemResponse> {
        let query = QueryClaimToSystemRequest {
            claim_type,
            trust_root: protobuf::MessageField::some(trust_root.clone()),
            query: Some(polycentric_protocol::protocol::query_claim_to_system_request::Query::MatchAnyField(match_any_field.to_string())),
            ..Default::default()
        };
        let encoded_query =
            base64::encode_config(query.write_to_bytes()?, base64::URL_SAFE);
        let url = format!("{}/resolve_claim?query={}", server, encoded_query);
        let request = self.get_request_builder(&url);

        let response = request.send().await?;
        Self::check_response("getResolveClaim", &response).await?;

        let query_response = QueryClaimToSystemResponse::parse_from_bytes(
            &response.bytes().await?,
        )?;
        Ok(query_response)
    }

    pub async fn get_query_latest(
        &self,
        server: &str,
        system: &PublicKey,
        event_types: &[u64],
    ) -> Result<Events> {
        let system_query =
            base64::encode_config(system.write_to_bytes()?, base64::URL_SAFE);
        let event_types_proto = RepeatedUInt64 {
            numbers: event_types.to_vec(),
            ..Default::default()
        };
        let event_types_query = base64::encode_config(
            event_types_proto.write_to_bytes()?,
            base64::URL_SAFE,
        );
        let url = format!(
            "{}/query_latest?system={}&event_types={}",
            server, system_query, event_types_query
        );
        let request = self.get_request_builder(&url);

        let response = request.send().await?;
        Self::check_response("getQueryLatest", &response).await?;

        let events = Events::parse_from_bytes(&response.bytes().await?)?;
        Ok(events)
    }

    pub async fn get_query_index(
        &self,
        server: &str,
        system: &PublicKey,
        content_type: u64,
        after: Option<i64>,
        limit: Option<i64>,
    ) -> Result<QueryIndexResponse> {
        let system_query =
            base64::encode_config(system.write_to_bytes()?, base64::URL_SAFE);
        let mut url = format!(
            "{}/query_index?system={}&content_type={}",
            server, system_query, content_type
        );

        if let Some(after_val) = after {
            url.push_str(&format!("&after={}", after_val));
        }
        if let Some(limit_val) = limit {
            url.push_str(&format!("&limit={}", limit_val));
        }

        let request = self.get_request_builder(&url);

        let response = request.send().await?;
        Self::check_response("getQueryIndex", &response).await?;

        let query_index_response =
            QueryIndexResponse::parse_from_bytes(&response.bytes().await?)?;
        Ok(query_index_response)
    }

    #[allow(clippy::too_many_arguments)]
    pub async fn get_query_references(
        &self,
        server: &str,
        reference: &Reference,
        cursor: Option<&[u8]>,
        request_events: Option<&QueryReferencesRequestEvents>,
        count_lww_element_references: Option<
            &[QueryReferencesRequestCountLWWElementReferences],
        >,
        count_references: Option<&[QueryReferencesRequestCountReferences]>,
        extra_byte_references: Option<&[Vec<u8>]>,
    ) -> Result<QueryReferencesResponse> {
        let mut query = QueryReferencesRequest {
            reference: protobuf::MessageField::some(reference.clone()),
            ..Default::default()
        };

        if let Some(c) = cursor {
            query.cursor = Some(c.to_vec());
        }
        if let Some(re) = request_events {
            query.request_events = protobuf::MessageField::some(re.clone());
        }
        if let Some(clr) = count_lww_element_references {
            query.count_lww_element_references = clr.to_vec();
        }
        if let Some(cr) = count_references {
            query.count_references = cr.to_vec();
        }
        if let Some(ebr) = extra_byte_references {
            query.extra_byte_references = ebr.to_vec();
        }

        let encoded_query =
            base64::encode_config(query.write_to_bytes()?, base64::URL_SAFE);
        let url =
            format!("{}/query_references?query={}", server, encoded_query);
        let request = self.get_request_builder(&url);

        let response = request.send().await?;
        Self::check_response("getQueryReferences", &response).await?;

        let query_references_response =
            QueryReferencesResponse::parse_from_bytes(
                &response.bytes().await?,
            )?;
        Ok(query_references_response)
    }

    pub async fn get_search(
        &self,
        server: &str,
        search_query: &str,
        limit: Option<u32>,
        cursor: Option<&[u8]>,
        search_type: Option<&str>,
    ) -> Result<Events> {
        // TODO: Check if search_query works right here
        let mut url = format!(
            "{}/search?search={}",
            server,
            base64::encode_config(search_query.as_bytes(), base64::URL_SAFE)
        );

        if let Some(c) = cursor {
            url.push_str(&format!(
                "&cursor={}",
                base64::encode_config(c, base64::URL_SAFE)
            ));
        }
        if let Some(l) = limit {
            url.push_str(&format!("&limit={}", l));
        }
        if let Some(st) = search_type {
            url.push_str(&format!("&search_type={}", st));
        }

        let response = self.get_request_builder(&url).send().await?;
        Self::check_response("getSearch", &response).await?;

        let events = Events::parse_from_bytes(&response.bytes().await?)?;
        Ok(events)
    }

    pub async fn get_top_string_references(
        &self,
        server: &str,
        query: Option<&str>,
        time_range: Option<&str>,
        limit: Option<u32>,
    ) -> Result<Events> {
        let mut url = format!("{}/top_string_references?", server);

        if let Some(q) = query {
            url.push_str(&format!("query={}&", q));
        }
        if let Some(tr) = time_range {
            url.push_str(&format!("time_range={}&", tr));
        }
        if let Some(l) = limit {
            url.push_str(&format!("limit={}&", l));
        }

        let response = self.get_request_builder(&url).send().await?;
        Self::check_response("getTopStringReferences", &response).await?;

        let events = Events::parse_from_bytes(&response.bytes().await?)?;
        Ok(events)
    }

    pub async fn get_head(
        &self,
        server: &str,
        system: &PublicKey,
    ) -> Result<Events> {
        let system_query =
            base64::encode_config(system.write_to_bytes()?, base64::URL_SAFE);
        let url = format!("{}/head?system={}", server, system_query);

        let response = self.get_request_builder(&url).send().await?;
        Self::check_response("getHead", &response).await?;

        let events = Events::parse_from_bytes(&response.bytes().await?)?;
        Ok(events)
    }

    pub async fn get_explore(
        &self,
        server: &str,
        limit: Option<u32>,
        cursor: Option<&[u8]>,
    ) -> Result<Events> {
        let mut url = format!("{}/explore?", server);

        if let Some(c) = cursor {
            url.push_str(&format!(
                "cursor={}&",
                base64::encode_config(c, base64::URL_SAFE)
            ));
        }
        if let Some(l) = limit {
            url.push_str(&format!("limit={}&", l));
        }

        let response = self.get_request_builder(&url).send().await?;
        Self::check_response("getExplore", &response).await?;

        let events = Events::parse_from_bytes(&response.bytes().await?)?;
        Ok(events)
    }

    pub async fn get_find_claim_and_vouch(
        &self,
        server: &str,
        vouching_system: &PublicKey,
        claiming_system: &PublicKey,
        fields: &[ClaimFieldEntry],
        claim_type: u64,
    ) -> Result<Option<FindClaimAndVouchResponse>> {
        let query = FindClaimAndVouchRequest {
            vouching_system: protobuf::MessageField::some(
                vouching_system.clone(),
            ),
            claiming_system: protobuf::MessageField::some(
                claiming_system.clone(),
            ),
            fields: fields.to_vec(),
            claim_type,
            ..Default::default()
        };

        let encoded_query =
            base64::encode_config(query.write_to_bytes()?, base64::URL_SAFE);
        let url =
            format!("{}/find_claim_and_vouch?query={}", server, encoded_query);

        let response = self.get_request_builder(&url).send().await?;

        if response.status() == reqwest::StatusCode::NOT_FOUND {
            return Ok(None);
        }

        Self::check_response("getFindClaimAndVouch", &response).await?;

        let find_claim_and_vouch_response =
            FindClaimAndVouchResponse::parse_from_bytes(
                &response.bytes().await?,
            )?;
        Ok(Some(find_claim_and_vouch_response))
    }

    pub async fn get_challenge(
        &self,
        server: &str,
    ) -> Result<HarborChallengeResponse> {
        let response = self
            .get_request_builder(&format!("{}/challenge", server))
            .send()
            .await?;
        Self::check_response("getChallenge", &response).await?;

        let challenge_response = HarborChallengeResponse::parse_from_bytes(
            &response.bytes().await?,
        )?;
        Ok(challenge_response)
    }

    pub async fn post_purge(
        &self,
        server: &str,
        solved_challenge: HarborValidateRequest,
    ) -> Result<()> {
        let response = self
            .post_request_builder(&format!("{}/purge", server))
            .body(solved_challenge.write_to_bytes()?)
            .send()
            .await?;

        Self::check_response("postPurge", &response).await
    }

    pub async fn post_claim_handle(
        &self,
        server: &str,
        claim_request: ClaimHandleRequest,
    ) -> Result<()> {
        let response = self
            .post_request_builder(&format!("{}/claim_handle", server))
            .body(claim_request.write_to_bytes()?)
            .send()
            .await?;

        Self::check_response("postClaimHandle", &response).await
    }

    pub async fn get_resolve_handle(
        &self,
        server: &str,
        handle: &str,
    ) -> Result<PublicKey> {
        let response = self
            .get_request_builder(&format!(
                "{}/resolve_handle?handle={}",
                server, handle
            ))
            .send()
            .await?;

        Self::check_response("getResolveHandle", &response).await?;

        let public_key = PublicKey::parse_from_bytes(&response.bytes().await?)?;
        Ok(public_key)
    }
}
